# Exercice
1. a) Créer une fonction "somme" qui prend en paramètre deux nombres a et b et affiche la somme de ces deux nombres.     
b) Créer la variable c égale à la somme des nombres 113 et 341.     
c) Calculer alors le carré de c.     
d) Que se passe-t-il ?     
2. a) Modifier la fonction "somme" qui prend en paramètre deux nombres a et b et retourne la somme de ces deux nombres.       
b) Créer la variable c égale à la somme des nombres 113 et 341.        
c) Calculer alors le carré de c.       
d) Que se passe-t-il ?

In [29]:
def somme(a,b):
    print(a+b)
    
c=somme(113,341)
c**2

454


TypeError: unsupported operand type(s) for ** or pow(): 'NoneType' and 'int'

In [30]:
def somme(a,b):
    return a+b
    
c=somme(113,341)
c**2

206116

# Exercice
1. Créer une fonction "somme_arithm" qui prend en paramètre un nombre entier N et retourne la somme de tous les nombres entiers inférieurs ou égaux à N.
2. Créer une fonction "somme_carre" qui prend en paramètre un nombre entier N et retourne la somme des carrés des nombres entiers inférieurs ou égaux à N.
3. Créer une fonction "somme_inverse" qui prend en paramètre un nombre entier N et retourne la somme des inverses des nombres entiers inférieurs ou égaux à N.
4. Créer une fonction "somme_inverse_carre" qui prend en paramètre un nombre entier N et retourne la somme des carrés des inverses des nombres entiers inférieurs ou égaux à N.

In [57]:
def somme_arithm(N):
    S=0
    for k in range(N):
        S=S+k
    return S

somme_arithm(10)

45

In [58]:
def somme_carre(N):
    S=0
    for k in range(N):
        S=S+k**2
    return S

somme_carre(10)

285

In [65]:
def somme_inverse(N):
    S=0
    for k in range(1,N):
        S=S+1/k
    return S

somme_inverse(1000)

7.484470860550343

In [69]:
def somme_inverse_carre(N):
    S=0
    for k in range(1,N):
        S=S+1/k**2
    return S

somme_inverse_carre(100000000)

1.644934057834575

# Exercice
1. Écrire une fonction "zen" qui affiche la ligne de 10 caractères : ZNZNZNZNZN
2. Modifier cette fonction pour prendre en **paramètre** le nombre n de caractères.
3. Modifier cette fonction pour prendre aussi en paramètre chaque caractère N et Z.
4. Utiliser la fonction zen pour créer les deux lignes suivantes :       
ZNZNZNZNZN    
NZNZNZNZNZ
5. Écrire une fonction "pluszen" qui utilise la fonction zen pour afficher p lignes et n colonnes de 2 caractères décalés.

In [1]:
def zen():
    for _ in range(5):
        print("ZN",end="")

In [2]:
zen()

ZNZNZNZNZN

In [3]:
def zen(n):
    for _ in range(n//2):
        print("ZN",end="")

In [4]:
zen(10)

ZNZNZNZNZN

In [11]:
def zen(n,c1,c2):
    for _ in range(n//2):
        print("%s%s" % (c1,c2),end="")
    print()

In [70]:
zen(10,"Z","N")
zen(10,"N","Z")

ZNZNZNZNZN
NZNZNZNZNZ


In [13]:
def pluszen(n,p):
    for _ in range(p//2):
        zen(n,"Z","N")
        zen(n,"N","Z")

In [16]:
pluszen(12,12)

ZNZNZNZNZNZN
NZNZNZNZNZNZ
ZNZNZNZNZNZN
NZNZNZNZNZNZ
ZNZNZNZNZNZN
NZNZNZNZNZNZ
ZNZNZNZNZNZN
NZNZNZNZNZNZ
ZNZNZNZNZNZN
NZNZNZNZNZNZ
ZNZNZNZNZNZN
NZNZNZNZNZNZ


# Exercice
1. Créer une fonction "est_pair" qui prend en paramètre un nombre entier et retourne True si le nombre est pair et False s'il ne l'est pas.
2. Créer une fonction "est_multiple" qui prend en paramètre 2 nombres entiers et qui retourne True si le premier nombre est un multiple du second et False sinon.
3. Un nombre entier est dit premier s'il est divisible par 1 et lui-même. Écrire une fonction "est_premier" qui prend en paramètre un nombre entier et retourne True s'il est premier et False sinon.

In [31]:
def est_pair(n):
    if n%2==0: return True
    else: return False

In [32]:
est_pair(9)

False

In [33]:
est_pair(10)

True

In [34]:
def est_multiple(n,p):
    if n%p==0: return True
    else: return False

In [35]:
est_multiple(12,3)

True

In [36]:
est_multiple(15,4)

False

In [37]:
def est_premier(n):
    for k in range(2,n):
        if n%k==0: return False
    return True

In [38]:
est_premier(10)

False

In [39]:
est_premier(21)

False

In [40]:
est_premier(17)

True

In [47]:
est_premier(15787)

True

# Exercice

On dit que deux nombres entiers A et B sont amicaux si la somme des diviseurs de A est égale à B et en même temps la somme des diviseurs de B est égale à A. Attention, on ne prend pas en compte A et B comme diviseur ([wikipedia - nombres amicaux](https://fr.wikipedia.org/wiki/Nombres_amicaux)).

1. Créer une fonction "somme\_diviseur" qui prend en paramètre un nombre entier N et retourne la somme des diviseurs de N (on ne doit pas prendre  en compte N comme diviseur).
2. Créer une fonction "nombres\_amicaux" qui prend en paramètre 2 nombres entiers A et B et retourne True s'ils sont amicaux et False sinon.
3. Vérifier que les nombres 220 et 284 sont amicaux. De même avec 12 285 et 14 595.

In [76]:
def somme_diviseur(N):
    S=0
    for k in range(1,N-1):
        if N%k==0:
            #print(k)
            S=S+k
    return S

somme_diviseur(21)

11

In [79]:
def nombres_amicaux(A,B):
    if somme_diviseur(A)==B and somme_diviseur(B)==A: return True
    else: return False
    
nombres_amicaux(12285,14595)

True

# Exercice
Créer une fonction "retourne" qui prend en paramètre 1 nombre entier et retourne le nombre écrit à l'envers. Par exemple, retourne(123)=321

In [54]:
def retourne(n):
    nb=''
    while n//10!=0:
        r=n%10
        nb=nb+str(r)
        n=n//10
    nb=nb+str(n)
    return int(nb)

In [56]:
retourne(1234567)

7654321